ts/user-picker/user-picker.controller.ts

```diff
@@ -0,0 +1,342 @@
+				if (newValue !== undefined) {
+					this.userPickerService.getUsersByIds(newValue).then((bypassedUsers: IUserLookup[]) => {
+						this.tidyUp(bypassedUsers).then((completedByPassedUsers: IUserLookup[]) => {
+							_.each(completedByPassedUsers, (byPassedUser: IUserLookup) => {
+								if (_.find(this.$scope.users, (user: IUserLookup) => { return user.id === byPassedUser.id; }) === undefined) {
+									this.$scope.users.push(byPassedUser);
+								}
+							});
+						});
+					});
+				}
+			});
+
+			this.$scope.$watchGroup(["appId", "operations"], (newValue: [number, string[]], oldValue: [number, string[]]) => {
+				if (angular.isDefined(newValue) && angular.isDefined(newValue[0]) &&
+					angular.isDefined(newValue[1]) && newValue[1].length > 0 &&
+					newValue[0] != oldValue[0] && !_.isEqual(newValue[1], oldValue[1])) {
+					this.resetUsers();
+					this.refresh();
+				}
+			});
+
+			this.$scope.find = (search: string): void => {
+				this.resetUsers();
+				this.refresh(search);
+			};
+
+			this.$scope.loadMore = (): void => {
+				if (this.$scope.displayAllUsers === undefined || this.$scope.displayAllUsers === false) {
+					this.$scope.lastPagingOffset += MAGIC_PAGING;
+					this.$scope.loadingMore = true;
+					this.refresh().then(() => { this.$scope.loadingMore = false; });
+				}
+			};
+
+			this.$scope.onSelectedUserChanged = (user: IUserLookup): void => {
+				this.setViewValue(user);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUsersChanged = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUserRemoved = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onRemove()) {
+					this.$scope.onRemove();
+				}
+			};
+		}
+
+		/**
+		 * Fetches most of the additional properties given in the attributes of the directive (custom-info, customInfoAsync, etc).
+		 * Also handles homonyms.
+		 */
+		private tidyUp(users: IUserLookup[]): ng.IPromise<IUserLookup[]> {
+			let promises = new Array<ng.IPromise<any>>();
+			let customInfoDico: { [userId: number]: ng.IPromise<string> } = {};
+			let homonymsDico = {};
+
+			_.each(users, (user: IUserLookup) => {
+				user.hasLeft = !!user.dtContractEnd && moment(user.dtContractEnd).isBefore(moment().startOf("day"));
+			});
+
+			if (this.$scope.customFilter) {
+				users = _.filter(users, (user: IUserLookup) => { return this.$scope.customFilter(user); });
+			}
+			if (!!this.$scope.customInfo) {
+				_.each(users, (user: IUserLookup) => {
+					user.info = this.$scope.customInfo(user);
+				});
+			}
+			if (!!this.$scope.customInfoAsync) {
+				_.each(users, (user: IUserLookup) => {
+					customInfoDico[user.id.toString()] = promises.push(this.$scope.customInfoAsync(user)) - 1;
```

Comment:
What I'm actually doing here is adding a property to the 'customInfoDico' object, i.e. customInfoDico["42"] = [...] : customInfoDico[42] wouldn't work since 'customInfoDico' isn't an array